### PR TITLE
Add AST visitor 

### PR DESCRIPTION
**Pull request by jpschorr**
Add AST visitor

- Add `Visit` and `Visitor` Traits for visiting AST
- Add AST node `Visit` impls via `proc_macro`s
- Refactor some AST nodes (including `FROM`,`WHERE`, and `HAVING` clauses) for better visitation

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-ast/src/ast.rs [hunk 1]
```diff
@@ -64,33 +66,38 @@ pub enum DdlOp {
     DropIndex(DropIndex),
 }
 
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Visit, Clone, Debug, PartialEq, Eq)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct CreateTable {
+    #[visit(skip)]
```
Comment (alancai98) [inline partiql-ast/src/ast.rs:72]: Out of curiosity, why are some `ast` nodes skipped?

File: partiql-ast/partiql-ast-macros/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    );
+    let exit_fn_name = Ident::new(
+        &format!("exit_{}", visit_fn_name),
+        proc_macro2::Span::call_site(),
+    );
+
+    let visit_children = (!should_skip_recurse(ast)).then(|| impl_visit_children(&ast));
+
+    let ast_name = &ast.ident;
+    quote! {
```
Comment (alancai98) [inline partiql-ast/partiql-ast-macros/src/lib.rs:64]: I'm still new to procedure macros in Rust. Is there a way to see what the procedural macro generates in `cargo` or CLion?

**Review by alancai98 — APPROVED** [review]
Looks good. Left a nit and a couple non-blocking questions

File: partiql-ast/src/ast.rs [hunk 3]
```diff
@@ -262,16 +427,30 @@ pub enum CollectionLit {
...
     pub qualifier: ScopeQualifier,
 }
 
-#[derive(Clone, Debug, PartialEq)]
+/// Indicates scope search order when resolving variables.
+/// Has no effect except within `FROM` sources.
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum ScopeQualifier {
+    /// Use the default search order.
```
Comment (am357) [inline partiql-ast/src/ast.rs:444]: Perhaps adding more info (or spec. reference to default order would be good).

**Review by am357 — APPROVED** [review]

File: partiql-ast/partiql-ast-macros/src/lib.rs (hunk 2 above)
Comment (jpschorr) [reply partiql-ast/partiql-ast-macros/src/lib.rs:64]: `cargo expand` in the directory of the crate using the macro will expand *all* macro usages.

File: partiql-ast/src/ast.rs (hunk 1 above)
Comment (jpschorr) [reply partiql-ast/src/ast.rs:72]: Importantly, it's not really 'ast nodes' that are skipped, but rather data items that inhere to the current node.

In some sense, we recursively visit 'children'. Things that aren't 'children' get 'skip'ed by the visitor macros, as it wouldn't really make sense for your visitor to have a `visit_symbol_primitive` method that is divorced from what node that symbol is appearing in.

**Pull request merged**